<a href="https://colab.research.google.com/github/JoelJ77/nitda-blockchain-scholarship/blob/main/Basic_Python_loops_examples.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/Python-Notebook-Banners/Examples.png"  style="display: block; margin-left: auto; margin-right: auto;";/>
</div>

# Examples: Basic Python loops


This notebook will guide us through how loops work in Python. In this lesson, we're going to cover the basics of `for` and `while` loops.


## Learning objectives

In this train, we will learn:

- **The importance of loops:** Grasp why loops are essential in programming for automating repetitive tasks and handling large datasets.

- **To differentiate between for and while loops:** Identify the scenarios in which each type of loop is most effective and understand their unique characteristics.

## The practical need for loops
Loops are a fundamental concept in programming, serving a crucial role in automating repetitive tasks and efficiently handling large datasets. Imagine we have a task, like checking whether an animal in our dataset is on an endangered species list.

Here we have a list of `animals` and we want to print them out one by one. We could use list indexing and print them out, one by one:

In [1]:
# List of animals (both endangered and not)
animals = ['Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala']

print(animals[0])
print(animals[1])
print(animals[2])
print(animals[3])
# etc.

Great White Shark
Blue Whale
African Elephant
Bald Eagle


But imagine our list contains 100 animals. We would have to write a print statement for each of them. This is a very repetitive task, and we are likely to make a mistake somewhere.

But loops were designed to solve this problem.

Python offers two primary types of loops: `for` loops and `while` loops.
Each serves different purposes:

* **`for` loops:** Ideal for going through a sequence (like a list, tuple, or string), one by one. We use `for` loops when we know how many things there are in a list.

    
    For example, for each element `i` in `list_of_numbers`, print out the value of `i`:

In [2]:
list_of_numbers =[0,1,2,3,4,5]
for i in list_of_numbers:
    # Loop body
    print(i)

0
1
2
3
4
5


* **`while` loops:** While loops continue to execute a set of instructions as long as a certain condition is met.
    
    For example, while `i` is smaller than `5`, print out the value of `i` and add 1 to it.

In [3]:
i = 0

while i < 5:
    print(i)
    i += 1

0
1
2
3
4


When we talk about loops, there is an important term to know:
* *Iterate*: To "iterate" in programming means to repeatedly execute a set of instructions. We can *iterate* through a list, or say in the last *iteration* of the while loop, i is assigned a value of 5.


## For loops
### Syntax and structure

The syntax of a for loop in Python is straightforward. It includes the `for` keyword, an `item` variable, the `in` operator, and a `collection` object, where `collection` refers to an object we **can** iterate through, like a list, dictionary, or tuple. The instructions we want to run in each loop are in the loop body.

In [2]:
# The syntax of a for loop in Python is straightforward. It includes the `for` keyword, an `item` variable, the `in` operator, and a `collection` object, where `collection` refers to an object we **can** iterate through, like a list, dictionary, or tuple. The instructions we want to run in each loop are in the loop body.
# for item in collection:
#     # Loop body

# Example using the animals list
for animal in animals:
    print(animal)

NameError: name 'animals' is not defined

For example:

Here the loop accesses each value in the list, one by one. We can iterate through our list of animals too, one by one:

In [3]:
for animal in animals:
    print(animal)

NameError: name 'animals' is not defined

We can add as many instructions as we like in the body of the loop. Once a loop is complete, it executes the next line of code after the loop.

In [4]:
animals_we_have_seen = [] #This is an empty list of animals we have encountered

for animal in animals:
    print(animal)
    animals_we_have_seen.append(animal) #add the animal to the "seen" list.

print(f" We saw these animals:\n {animals_we_have_seen}")

NameError: name 'animals' is not defined

We can loop through any iterable data structure like tuples, sets, dictionaries, and even strings:

In [6]:
animals_tuple = ('Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Tiger', 'Panda', 'Koala')
animals_set= set(animals_tuple)
animal_string = "Great White Shark"

In [11]:
print("Tuple:")
for animal in animals_tuple:
    print(animal)

Tuple:
Great White Shark
Blue Whale
African Elephant
Bald Eagle
Orangutan
Tiger
Tiger
Panda
Koala


In [12]:
print("Set:")
for animal in animals_set:
    print(animal)

Set:
Koala
Panda
Bald Eagle
Tiger
Orangutan
Blue Whale
African Elephant
Great White Shark


In [13]:

print("String:")
for letter in animal_string:
    print(letter)

String:
G
r
e
a
t
 
W
h
i
t
e
 
S
h
a
r
k


## While loops

A `while` loop in Python repeatedly executes its body **as long as a condition is true**. Its syntax is:

In [ ]:
while condition:
    # Loop body

`while` loops are ideal for scenarios where you don't know how many times you need to iterate. For example, adding 10% to a value, until it is > 5. It is hard to know before we start how many times we have to repeat the calculation, so if we use a `while` loop, we can keep going until we get to the answer.

In [16]:
population = 1 #We start at 0

while population < 3:
    print(population) # We print out the value of population
    population *= 1.1 # And add 10% to it.

1
1.1
1.2100000000000002
1.3310000000000004
1.4641000000000006
1.6105100000000008
1.771561000000001
1.9487171000000014
2.1435888100000016
2.357947691000002
2.5937424601000023
2.853116706110003


This loop prints the population with 10% added to each iteration, until the population is 2.85, and the loop stops because the next number is larger than 3.

Be cautious with `while` loops, because they result in infinite loops if the condition never becomes false. Always ensure there's a way for the condition to become false eventually.

If your loop is infinitely looping, press the `i` key twice rapidly, or click on the stop/interrupt button to interrupt the kernel. The cell below contains an example of an infinite loop. It will go into an infinite loop when running it, so beware!

In [15]:
count = 0
while count >= 0:  # This condition is always true
    print("This will go on forever!")

Streaming output truncated to the last 5000 lines.
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!
This will go on forever!

KeyboardInterrupt: 

## Loop controls
Loop control statements are used to change the execution flow of loops. They provide flexibility and control over how loops behave in different scenarios. The primary loop control statements in Python are `break`, `continue`, and `pass`.

### `break` statement
`break` immediately stops the execution of a loop, and the control flows to the statement immediately after the loop.

`break` is often used when searching for an item in a list. Once the item is found, there's no need to continue the loop.

In [17]:
# Searching for a specific number in a list
animals = ['Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala']

for animal in animals:
    print(animal)
    if animal == 'African Elephant':
        print("African Elephant found")
        break #Once African Elephant is found, the loop stops, so 'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala' are not printed.

print(f"The last animal checked = {animal}")

Great White Shark
Blue Whale
African Elephant
African Elephant found
The last animal checked = African Elephant


The loop stopped once we reached the `African Elephant` animal, so printing out `print(f"The last animal checked = {animal}")` after the loop has completed, results in `African Elephant`, showing that the loop stopped there.

### `continue` statement
`continue` skips the rest of the loop's **body** for the current iteration and moves to the next iteration.

This is useful when we need to skip certain elements in a loop, like filtering out specific values.

In [18]:
for animal in animals:

    if animal == 'African Elephant':
        continue #Once African Elephant is found, the code in the rest of the loop is not executed, so African Elephant found is not printed.
    print(animal)

Great White Shark
Blue Whale
Bald Eagle
Orangutan
Tiger
Panda
Koala


### `pass` statement
`pass` is a statement that does nothing. It's used as a placeholder.

`pass` is used when we need a statement to have the correct syntax but want to implement no action, like in empty definitions, loops, or conditionals. `pass` is very useful as a placeholder for the code we are going to add later.

Suppose we want to check all animals, but skip `African Elephants`:

In [19]:
#This code will produce an error because it is expecting code in the if-statement.
for animal in animals:
    if animal == 'African Elephant':
        #Do Nothing
    else:
        print (f"{animal} is not an African Elephant")

IndentationError: expected an indented block after 'if' statement on line 3 (ipython-input-3319311374.py, line 5)

In [20]:
# By adding pass, the code still runs
for animal in animals:
    if animal == 'African Elephant':
        pass #Skip elephants
    else:
        print (f"{animal} is not an African Elephant")

Great White Shark is not an African Elephant
Blue Whale is not an African Elephant
Bald Eagle is not an African Elephant
Orangutan is not an African Elephant
Tiger is not an African Elephant
Panda is not an African Elephant
Koala is not an African Elephant


Be careful of using `pass` because it often tells us our logic can be improved. Instead of specifying that we do nothing when the animal is an African Elephant with `pass`, we can change the conditional flow logic:

In [21]:
# By improving our logic, we avoid the use of pass.
for animal in animals:
    if animal != 'African Elephant': #If the animal is not an elephant, then...
        print (f"{animal} is not an African Elephant")

Great White Shark is not an African Elephant
Blue Whale is not an African Elephant
Bald Eagle is not an African Elephant
Orangutan is not an African Elephant
Tiger is not an African Elephant
Panda is not an African Elephant
Koala is not an African Elephant


If we have to use `pass` in our code, it is an indication we can improve our logic somewhere.

## Summary
In this notebook, we've explored the fundamental concepts of loops in Python, including their syntax, different types, and practical applications. We've learned that **loops are essential** for automating **repetitive** tasks and efficiently processing collections of data. By understanding the nuances of `for` and `while` loops, as well as loop control statements like `break`, `continue`, and `pass`, we can write more effective and efficient Python code.


#  

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/refs/heads/master/ALX_banners/ALX_Navy.png"  style="width:140px";/>
</div>

In [ ]:
# List of animals (both endangered and not)
animals = ['Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala']